### ISSUE TITLE
Procedures marked with an explicit 'ref' intent that do not return anything will ignore 'halt' and 'compilerError' and whether or not it is unused.

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#10936]
Normally when you have a function that explicitly states that it returns something but does not, such as for 'place-holder' methods, a `halt` or `compilerError` is normally enough to prevent compiler errors. That is the following is fine and does not generate a compiler error...

```chpl
proc p() : int {
   halt();
}
```

or

```chpl
proc p() : int {
   compilerError();
}
```

As well, something like this okay too because the compiler knows that the function is never actually used...

```chpl
proc p() : int {

}
```

However, once an explicit `ref` is attached to the end of it, it will crash in all 3 of the above cases.

```chpl
proc p() ref {

}
```
[TIO](https://tio.run/##S85ILEjN@f@/oCg/WaFAQ1OhKDVNobqWi@v/fwA)

My assumption is that `ref` return types are inferred extremely early, even earlier than dead-code elimination.

**Comment by bradcray** [comment]
If I were to try and restate your request, is the following correct:

"I'd like to be able to declare a function/method that returns by `ref` intent, yet which doesn't return anything, by putting `halt()` or `compilerError()` into its body"?

> As well, something like this okay too because the compiler knows that the function is never actually used...

Since over in issue #10449 we're talking about resolving all concrete functions whether or not they're called, I don't think we should expend effort making this case work at present (because as soon as that issue is done, it won't).  I agree with your point that the current behavior points to refs being handled too early.

In a quick look, it appears to me that grepping the compiler sources for the error in question and removing that conditional from the compiler makes things behave as you'd like.  I haven't run testing on it.

**Comment by bradcray** [comment]
I ran my simple change through testing and it failed (surprise, surprise) on only the one test that was making sure we generated an error for this case (when the function was called).  Later passes of the compiler don't complain when a (called) function marked `ref` / `const ref` doesn't return anything, but I think they should, so am looking into whether that can be enabled without too much effort.

**Comment by lydia-duncan** [comment]
@LouisJenkinsCS - note that @DanilaFe resolved this recently with his #25494.  Thanks for reporting, and thanks for the fix, Daniel!
